src/coprocessor/mod.rs

```diff
@@ -78,6 +78,7 @@ pub trait Coprocessor<F: LurkField>: Clone + Debug + Sync + Send + CoCircuit<F>
         input_expr: &AllocatedPtr<F>,
         input_env: &AllocatedPtr<F>,
         input_cont: &AllocatedContPtr<F>,
+        dummy_or_blank: bool,
```

Comment:
LEM's infra passes down a `not_dummy: &Boolean`. Would it be possible to use such kind of data here?

I think encoding the "not dummy" semantics is easy, but then the "blank" case would have to be dealt with in `circuit_frame.rs`.

This suggestion is meant to facilitate the API compatibility with LEM

Reply:
Done

Reply:
Well, I changed the sense and meaning generally, but still pass a `bool`. I *could* allocate a `Boolean` in order to pass it, but I'm concerned that this would blur the actual coprocessor contract. If LEM, for some reason *requires* and takes advantage of this, then coprocessors written to that API won't be equivalent, I think.

Reply:
That having been said, I think this may actually need to be allocated (and used as such) for correctness… Let me try to verify and fix if so.

Reply:
Nice discussion. Now I think that my motivation to downstream a `&Boolean` was not correct. I did so to allow the coprocessor author to make enforcements with implications (see `DumbCoprocessor`). But the tag check should return an error continuation instead of panicking.

I will fix it and later adequate it to the design you choose here.

Reply:
Yes, good discussion. It seems to have convinced me that we do actually need to do that here, though.

This coprocessor is hard to reason about, especially since it's the first non-trivial one mixing and matching a custom circuit with a Lurk circuit.

I think it was underconstrained before. In order to fully constrain the trie's root in the not-dummy case, I think we need the implication. So I suggest waiting until I've resolved this to do anything, since we might be going in opposite directions.

Reply:
Okay, so I won't change the interface because that will ultimately depend on the design you need here. But I will fix my misuse of the `&Boolean` because that's just not the semantic we want to encode in the `DumbCoprocessor`

Reply:
I've pushed a commit that indeed switches to use `Boolean` here. That was needed once I started properly constraining the trie root to be the one supplied by the call. The issue is that in a dummy case, such a constraint will fail unless the passed value *happens* to correspond to a valid trie the store knows about. So we need to only enforce the constraint in the `not_dummy` case.